File: 1100/1345B/main.cpp
#include <bits/stdc++.h>
using namespace std;

long long TotalCards(long long height) 
{
    // Function to calculate the total number of cards needed to build a pyramid of given height
    return (3LL * height * height + height) / 2;
}

int main() 
{
    int t;
    cin >> t;

    while (t--) 
    {
        long long n;
        cin >> n;

        int cnt = 0;
        while (n >= 2) // Minimum cards needed to build a pyramid of height 1 is 2
        {
            long long low = 1, high = 100000, height = 0;

            while (low <= high) // Binary search to find the maximum height of the pyramid that can be built with n cards
            {
                long long mid = low + (high - low) / 2;

                if (TotalCards(mid) <= n) // Check if we can build a pyramid of height mid with n cards
                {
                    height = mid;
                    low = mid + 1;
                } 
                else // If we cannot build a pyramid of height mid, try smaller heights
                    high = mid - 1;
            }

            n -= TotalCards(height); // Subtract the cards used to build the pyramid from n
            cnt++;
        }

        cout << cnt << "\n";
    }

    return 0;
}

File: 1100/1345B/main.java
import java.util.*;
public class main
{
    public static long TotalCards(int height)
    {
        return (3L*height*height + height)/2;
    }
    
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t-- > 0)
        {
            long n = sc.nextInt();
            
            int cnt = 0;
            while(n >= 2)
            {
                int low = 1;
                int high = 100000;
                int height = 0;
                
                while(low <= high)
                {
                    int mid = low + (high-low)/2;
                    
                    if(TotalCards(mid) <= n)
                    {
                        height = mid;
                        low = mid+1;
                    }
                    else
                        high = mid-1;
                }
                
                n -= TotalCards(height);
                cnt++;
            }
            
            System.out.println(cnt);
        }
    }
}